ta/trusted_key/entry.c

```diff
@@ -0,0 +1,308 @@
+	} else if (mode == TEE_MODE_DECRYPT) {
+		res = huk_AE_decrypt(crypto_op, in, in_sz, out, out_sz);
+		if (res != TEE_SUCCESS)
+			EMSG("huk_AE_decrypt failed: returned 0x%x\n", res);
+	}
+
+out_key:
+	TEE_FreeTransientObject(hkey);
+out_op:
+	TEE_FreeOperation(crypto_op);
+	memzero_explicit(huk_key, sizeof(huk_key));
+	return res;
+}
+
+static TEE_Result seal_trusted_key(uint32_t types,
+				   TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_SUCCESS;
+	size_t in_sz = 0, out_sz = 0;
+
+	FMSG("Invoked TA_CMD_SEAL");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	in_sz = params[0].memref.size;
+	out_sz = params[1].memref.size;
+
+	if (!params[0].memref.buffer || in_sz == 0 || in_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (!params[1].memref.buffer || out_sz <= (IV_SIZE + TAG_SIZE) ||
+	    out_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if ((in_sz + IV_SIZE + TAG_SIZE) > out_sz)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = huk_crypt(TEE_MODE_ENCRYPT, params[0].memref.buffer,
+			in_sz, params[1].memref.buffer, &out_sz);
+	if (res == TEE_SUCCESS)
+		params[1].memref.size = out_sz;
+	else
+		params[1].memref.size = 0;
+
+	return res;
+}
+
+static TEE_Result unseal_trusted_key(uint32_t types,
+				     TEE_Param params[TEE_NUM_PARAMS])
+{
+	TEE_Result res = TEE_SUCCESS;
+	size_t in_sz = 0, out_sz = 0;
+
+	FMSG("Invoked TA_CMD_UNSEAL");
+
+	if (types != TEE_PARAM_TYPES(TEE_PARAM_TYPE_MEMREF_INPUT,
+				     TEE_PARAM_TYPE_MEMREF_OUTPUT,
+				     TEE_PARAM_TYPE_NONE,
+				     TEE_PARAM_TYPE_NONE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	in_sz = params[0].memref.size;
+	out_sz = params[1].memref.size;
+
+	if (!params[0].memref.buffer || in_sz <= (IV_SIZE + TAG_SIZE) ||
+	    in_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+	if (!params[1].memref.buffer || out_sz == 0 || out_sz > MAX_BUF_SIZE)
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	if (in_sz > (out_sz + IV_SIZE + TAG_SIZE))
+		return TEE_ERROR_BAD_PARAMETERS;
+
+	res = huk_crypt(TEE_MODE_DECRYPT, params[0].memref.buffer,
+			in_sz, params[1].memref.buffer, &out_sz);
+	if (res == TEE_SUCCESS)
+		params[1].memref.size = out_sz;
+	else
+		params[1].memref.size = 0;
+
+	return res;
+}
+
+TEE_Result TA_CreateEntryPoint(void)
+{
+	return TEE_SUCCESS;
+}
+
+void TA_DestroyEntryPoint(void)
+{
+}
+
+TEE_Result TA_OpenSessionEntryPoint(uint32_t pt __unused,
+				    TEE_Param params[4] __unused,
+				    void **session __unused)
+{
+	TEE_Result res = TEE_ERROR_GENERIC;
+	TEE_PropSetHandle h = TEE_HANDLE_NULL;
+	TEE_Identity id = { 0 };
+
+	res = TEE_AllocatePropertyEnumerator(&h);
+	if (res != TEE_SUCCESS)
+		return res;
+
+	TEE_StartPropertyEnumerator(h, TEE_PROPSET_CURRENT_CLIENT);
+
+	res = TEE_GetPropertyAsIdentity(h, NULL, &id);
+	if (res != TEE_SUCCESS) {
+		TEE_FreePropertyEnumerator(h);
```

Comment:
`goto out:`?

Reply:
With separate `res_login` proposed [here](https://github.com/OP-TEE/optee_os/pull/3838#discussion_r420668406), I think we need to stick with original implementation.